### PR TITLE
JDK-6782021: It is not possible to read local computer certificates with the SunMSCAPI provider

### PR DESCRIPTION
**Pull request by macarte**
On Windows you can now access the local machine keystores using the strings "Windows-MY-LOCALMACHINE" and "Windows-ROOT-LOCALMACHINE"; note the application requires admin privileges.

"Windows-MY" and "Windows-ROOT" remain unchanged, however given these original keystore strings mapped to the current user, I added "Windows-MY-CURRENTUSER" and "Windows-ROOT-CURRENTUSER" so that a developer can explicitly specify the current user location. These two new strings simply map to the original two strings, i.e. no duplication of code paths etc

No new tests added, keystore functionality and API remains unchanged, the local machine keystore types would require the tests to run in admin mode

Tested on windows, passes tier1 and tier2 tests

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires a CSR request to be approved
- [x] Change must be properly reviewed (1 review required, with at least 1 reviewer)

### Issues
 * [JDK-6782021](https://bugs.openjdk.java.net/browse/JDK-6782021): It is not possible to read local computer certificates with the SunMSCAPI provider
 * [JDK-8284850](https://bugs.openjdk.java.net/browse/JDK-8284850): It is not possible to read local computer certificates with the SunMSCAPI provider (**CSR**)

### Reviewers
 * [Weijun Wang](https://openjdk.java.net/census#weijun) (@wangweij - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/8211/head:pull/8211` \
`$ git checkout pull/8211`

Update a local copy of the PR: \
`$ git checkout pull/8211` \
`$ git pull https://git.openjdk.java.net/jdk pull/8211/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 8211`

View PR using the GUI difftool: \
`$ git pr show -t 8211`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/8211.diff">https://git.openjdk.java.net/jdk/pull/8211.diff</a>

</details>

### CODE REVIEW

**Comment by wangweij** [comment]
Thanks a lot for the quick contribution! I'll definitely be happy to sponsor it.

Since new keystore types are created, this needs a CSR. Go to https://bugs.openjdk.java.net/browse/JDK-6782021, click "More" and at the end there is a "Create CSR". Basically you talk about what these new keystores are and why they are useful. The scope is JDK and I assume the compatibility risk is low. There is no spec change but you can suggest new entries in the "JDK Providers Documentation" (https://docs.oracle.com/en/java/javase/18/security/oracle-providers.html#GUID-4F1737D6-1569-4340-B140-678C70E63CD5). You can also add a label like `noreg-other` to the bug and add a comment explaining what manual tests can be added, how to run it, and what the expected output is.

**Comment by mkarg** [comment]
Actually the Windows API is using a reverse naming system, so I would propose the following changes:
* MY and ROOT become marked as deprecated in both, the source code and all documentation, so in future there won't be any confusion anymore what store these actually do access.
* Following the Windows naming system and make the syntax more logical, I propose the new store names are `Windows:LOCALMACHINE/MY` , `Windows:LOCALMACHINE/ROOT`, `Windows:CURRENTUSER/MY` and `Windows:CURRENTUSER/ROOT`. The colon is commonly well-understood as a separator between schemes and names, and the slash is commonly understood as a path separator. So this new syntax makes usage much cleaner.

**Comment by macarte** [comment]
Please see draft CSR: https://bugs.openjdk.java.net/browse/JDK-8284850

File: src/jdk.crypto.mscapi/windows/native/libsunmscapi/security.cpp
Comment (ecki) [inline src/jdk.crypto.mscapi/windows/native/libsunmscapi/security.cpp:419]: Would it be easier to use a int argument here, it’s a internal api with only two enumerated Types. That safes the jstring conversion and the Strcmp.

File: src/jdk.crypto.mscapi/windows/native/libsunmscapi/security.cpp
Comment (ecki) [inline src/jdk.crypto.mscapi/windows/native/libsunmscapi/security.cpp:487]: It is not quite clear from the CSR, was this a bug (previous JDK-8026953 incomplete) or is that only a type cleanup and using ncrypt keys worked before? (In that Case does it need to be mentioned in csr?)

File: src/jdk.crypto.mscapi/windows/native/libsunmscapi/security.cpp
Comment (ecki) [reply src/jdk.crypto.mscapi/windows/native/libsunmscapi/security.cpp:419]: And also, is there a ReleaseString missing?

**Comment by wangweij** [comment]
The CSR looks fine. It's not necessary to attach a patch there. At least don't name it `specdiff` because it's not about spec.

BTW, how do you think of the names proposed in https://github.com/openjdk/jdk/pull/8211#issuecomment-1099925744?

File: src/jdk.crypto.mscapi/windows/native/libsunmscapi/security.cpp
Comment (wangweij) [reply src/jdk.crypto.mscapi/windows/native/libsunmscapi/security.cpp:487]: Same question. Does a new type name automagically add support for CNG?

File: src/jdk.crypto.mscapi/windows/native/libsunmscapi/security.cpp
Comment (macarte) [reply src/jdk.crypto.mscapi/windows/native/libsunmscapi/security.cpp:419]: Thanks for the feedback, I'm going to incorporate that into the PR

File: src/jdk.crypto.mscapi/windows/native/libsunmscapi/security.cpp
Comment (macarte) [reply src/jdk.crypto.mscapi/windows/native/libsunmscapi/security.cpp:487]: Correct, it does enable access to certificates and keys that require next (second) generation, that were previously inaccessible.  I've just realized the implication of this on existing applications and so I'm going to pause and run some test, especially around enumeration

File: src/jdk.crypto.mscapi/windows/native/libsunmscapi/security.cpp
Comment (macarte) [reply src/jdk.crypto.mscapi/windows/native/libsunmscapi/security.cpp:419]: > And also, is there a ReleaseString missing?

Yes an error when I "patched" my repo, but based on the feedback there will no longer be a string to free :)

File: src/jdk.crypto.mscapi/windows/native/libsunmscapi/security.cpp
Comment (macarte) [reply src/jdk.crypto.mscapi/windows/native/libsunmscapi/security.cpp:487]: Correction: after looking at wincrypt.h, the documentation [1] and running tests, I can confirm that:
1) this change has no functional impact (i.e. results are unchanged)
2) HCRYPTPROV and HCRYPTPROV_OR_NCRYPT_KEY_HANDLE are both the same type (ULONG_PTR) and so are interchangeable (with the former supporting legacy applications)
3) There is only one function for CryptAcquireCertificatePrivateKey, not two differentiated by this parameter type change
4) support for CNG keys, which was originally thought to be added with this change, has always been true due to the existing use of the flag CRYPT_ACQUIRE_ALLOW_NCRYPT_KEY_FLAG

I think this change should stay as it more correctly matches the prototype and the use of CRYPT_ACQUIRE_ALLOW_NCRYPT_KEY_FLAG 

[1] https://docs.microsoft.com/en-us/windows/win32/api/wincrypt/nf-wincrypt-cryptacquirecertificateprivatekey

File: src/jdk.crypto.mscapi/windows/native/libsunmscapi/security.cpp
Comment (macarte) [reply src/jdk.crypto.mscapi/windows/native/libsunmscapi/security.cpp]: Committed changes to use ints instead of strings

**Comment by macarte** [comment]
I don't use this API much so I don't really have an opinion as a customer regarding the format of the strings used to identify the key stores.   I'd be happy to review a separate PR but I think this falls outside the scope of this PR which specifically targets the inability to access local machine key stores (which a bug has been raised against).

note: there's also a "Windows-PRNG" which isn't a key store but the native random number generator

File: src/jdk.crypto.mscapi/windows/classes/sun/security/mscapi/CKeyStore.java
Comment (wangweij) [inline src/jdk.crypto.mscapi/windows/classes/sun/security/mscapi/CKeyStore.java:256]: Why not just an `int` here? The creation of a separate class `keyStoreLocation` seems not necessary. If you want code to be readable, just add `public static final int CURRENTUSER = 0`, etc.

**Comment by wangweij** [comment]
@macarte I've added my name as reviewer in the CSR. You can either propose it (if you expect some feedback from the CSR reviewers) or just finalize it (if you think the change is simple and obvious).

**Comment by wangweij** [comment]
Also, please remove trailing spaces and create a new commit. Skara dislikes trailing spaces and TAB characters in source code.

**Comment by macarte** [comment]
@wangweij - I believe the change to be simple enough I'll go ahead and finalize it.  However, I've proposed updates to the documentation, how are these actioned, i.e. what steps are required of me?

**Comment by wangweij** [comment]
> @wangweij - I believe the change to be simple enough I'll go ahead and finalize it. However, I've proposed updates to the documentation, how are these actioned, i.e. what steps are required of me?

I filed a doc task at https://bugs.openjdk.java.net/browse/JDK-8286141. It will be picked up by the doc team. Someone will need to write a release note. If you have any recommended text, you can write here.

File: src/jdk.crypto.mscapi/windows/classes/sun/security/mscapi/CKeyStore.java
Comment (macarte) [reply src/jdk.crypto.mscapi/windows/classes/sun/security/mscapi/CKeyStore.java]: I was using type safety to remove the chance of non-expected values being passed to the C function.  Implemented your recommendation as its a simple contract between two files

**Comment by macarte** [comment]
Found an issue with my editor where whitespace was not being rendered :( - will check the diff in future

File: src/jdk.crypto.mscapi/windows/classes/sun/security/mscapi/CKeyStore.java
Comment (wangweij) [inline src/jdk.crypto.mscapi/windows/classes/sun/security/mscapi/CKeyStore.java:866]: We usually indent 8 spaces in a continuation line.

File: src/jdk.crypto.mscapi/windows/classes/sun/security/mscapi/CKeyStore.java
Comment (wangweij) [inline src/jdk.crypto.mscapi/windows/classes/sun/security/mscapi/CKeyStore.java:860]: Take this chance to modify `Load` to `Loads`.

**Comment by wangweij** [comment]
I'd like to contribute a test. You can put it inside `test/jdk/sun/security/mscapi/`.
```
/*
 * Copyright (c) 2022, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 */

import jdk.test.lib.Asserts;
import jdk.test.lib.SecurityTools;

import java.security.KeyStore;
import java.util.Collections;
import java.util.List;
import java.util.Locale;

/*
 * @test
 * @bug 6782021
 * @requires os.family == "windows"
 * @library /test/lib
 * @summary More keystore types
 */
public class AllTypes {
    public static void main(String[] args) throws Exception {
        var nm = test("windows-my");
        var nr = test("windows-root");
        var nmu = test("windows-my-currentuser");
        var nru = test("windows-root-currentuser");
        var nmm = test("windows-my-localmachine");
        var nrm = test("windows-root-localmachine");
        Asserts.assertEQ(nm, nmu);
        Asserts.assertEQ(nr, nru);
    }

    private static List<String> test(String type) throws Exception {
        var stdType = "Windows-" + type.substring(8).toUpperCase(Locale.ROOT);
        SecurityTools.keytool("-storetype " + type + " -list")
                .shouldHaveExitValue(0)
                .shouldContain("Keystore provider: SunMSCAPI")
                .shouldContain("Keystore type: " + stdType);
        KeyStore ks = KeyStore.getInstance(type);
        ks.load(null, null);
        var content = Collections.list(ks.aliases());
        Collections.sort(content);
        return content;
    }
}
```

**Comment by wangweij** [comment]
Oops, I didn't realized that. Yes, you can divide it to 2 tests. The one needs admin privilege can be tagged `@run main/manual Test`. It probably won't get a lot of chance to run.

**Comment by wangweij** [comment]
@macarte You need to finalize your CSR soon if you want to include this change into JDK 19. RDP1 is 2022/06/09, and all enhancements require approval after that.

BTW, is it possible to detect whether you have admin privilege inside the test? There is a `NTSystem` class inside Java which can find out user info. Maybe the `getGroupIDs()` method will tell you if you are an admin?

**Comment by wangweij** [comment]
Great. After the CSR is approved I will approve this PR as well. Then you will need to type the `/integrate` command as a comment here, and I will type `/sponsor`. Then the Skara bot will integrate the change and close the PR.

**Comment by wangweij** [comment]
@macarte Can you explain more on when an "access denied" error could happen? I create a normal user in my local AD and didn't see that. We might need to talk about this in the release note.

File: test/jdk/sun/security/mscapi/AllTypes.java
Comment (wangweij) [inline test/jdk/sun/security/mscapi/AllTypes.java:60]: Could the message be different if the system language is not English?

Comment (macarte) [reply test/jdk/sun/security/mscapi/AllTypes.java:60]: I checked the API and its not guaranteed to be English.  I'm investigating a better method for detecting admin rights

**Comment by christophbrejla** [comment]
Hello, as i understand the target version for this fix to be included is JDK 19 yes?  is there also a possibility to request and include this fix in different versions like JDK 8 ?

thank you

**Comment by macarte** [comment]
@christophbrejla - my goal is to backport to latest (18 or 19), 17 and 11

**Comment by wangweij** [comment]
> @christophbrejla - my goal is to backport to latest (18 or 19), 17 and 11

Then please add the versions to the "Fix Version(s)" field of the CSR. There are also some questions waiting for you in the comment there.

**Comment by wangweij** [comment]
@macarte I think Sean's comment on your CSR about the scope is correct. The "algorithm" name should be at the JDK level so user knows what to write in their app. Once you think there's no more to update, you can finalize the CSR.

**Review by wangweij — APPROVED** [review]

**Comment by christophbrejla** [comment]
Hello @macarte <https://github.com/macarte>

thanks for the info. I dont know how much work it would be, but if it is not too much of a hassle i would much appreciate if you also backport to version 8. 

thank you very much

**Pull request closed without merging**
